Wikipedia/Code/WebViewController.m

```diff
@@ -249,6 +262,142 @@ - (void)handleArticleStateScriptMessage:(NSString*)messageString {
     }
 }
 
+- (void)handleFindInPageMatchesFoundMessage:(NSArray*)messageArray {
+    self.findInPageMatches = messageArray;
+    self.findInPageResultCursorIndex = -1;
+}
+
+#pragma mark - Find-in-page
+
+- (FindInPageKeyboardBar *)findInPageKeyboardBar {
+    return self.view.inputAccessoryView;
+}
+
+- (UITextField *)findInPageTextField {
+    return [self findInPageKeyboardBar].textField;
+}
+
+- (BOOL)isFindInPageVisible {
+    return [[self findInPageTextField] isFirstResponder];
+}
+
+- (void)showFindInPage {
+    [self referencesHide];
+    [self.view becomeFirstResponder];
+    [[self findInPageTextField] becomeFirstResponder];
+}
+
+- (void)hideFindInPage {
+    [self clearFindInPage];
+    [[self findInPageTextField] resignFirstResponder];
+    [self.view resignFirstResponder];
+    [self.webView evaluateJavaScript:@"window.wmf.findInPage.removeSearchTermHighlights()"
+                   completionHandler:nil];
+}
+
+- (void)clearFindInPage {
+    self.findInPageMatches = @[];
+    self.findInPageResultCursorIndex = -1;
+    [[self findInPageTextField] setText:@""];
```

Comment:
Example of what I mentioned above. The web view controller is setting text on the textfield, but so does the input view. So it is touch to answer the question: "which object is responsible modifying the text of this text field?" Its always best to strive to make that a single object.